toke.c

```diff
@@ -13932,6 +13932,84 @@ Perl_parse_subsignature(pTHX_ U32 flags)
     return parse_recdescent_for_op(GRAMSUBSIGNATURE, LEX_FAKEEOF_NONEXPR);
 }
 
+/*
+=for apidoc valid_identifier_pvn
+
+Returns true if the string given by C<s> whose length is C<len> would be
+considered valid as a Perl identifier.  That is, it must begin with a
+character matching C<isIDFIRST>, followed by characters all matching
+C<isIDCONT>.  An empty string (i.e. when C<len> is zero) will return false.
+
+If C<flags> contains the C<SVf_UTF8> bit, then the string is presumed to be
+encoded in UTF-8, and suitable Unicode character test functions will be used.
+
+=cut
+*/
+
+bool
+Perl_valid_identifier_pvn(pTHX_ const char *s, STRLEN len, U32 flags)
+{
+    PERL_ARGS_ASSERT_VALID_IDENTIFIER_PVN;
+
+    if(flags & SVf_UTF8) {
+        const U8 *end = (U8 *)s + len;
+        if(!len || !isIDFIRST_utf8_safe((U8 *)s, end))
+            return false;
+
+        while(1) {
+            STRLEN skip = UTF8SKIP((U8 *)s);
+            if(skip > len)
+                return false;
+            s += skip; len -= skip;
+            if(!len)
+                break;
+            if(!isIDCONT_utf8_safe((U8 *)s, end))
+                return false;
+        }
```

Comment:
Couldn't this loop be more simply written as
```
while (s < end) {
    s += UTF8SKIP((U8 *) s);
    if (s >= end || ! isIDCONT_utf8_safe((U8 *) s, end)) {
        return false;
    }
}

```
And the second loop below similarly simplified?

Reply:
Not quite; it still needs the `if(s == end) break` condition, but it seems fine if I add that.

Updated and force-pushed.